File: 160-Intersection-of-Two-Linked-Lists.py
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        if not headA and not headB:
            return None

        lenA = lenB = 0
        a, b = headA, headB

        while a:
            a = a.next
            lenA += 1

        while b:
            b = b.next
            lenB += 1

        if lenA < lenB:
            for i in range(lenB - lenA):
                headB = headB.next
        else:
            for i in range(lenA - lenB):
                headA = headA.next

        while headA and headB:
            if headA == headB:
                return headA
            headA, headB = headA.next, headB.next

        return None

File: neetcode/160-Intersection-of-Two-Linked-Lists.ts
function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
  if (!headA || !headB) return null;

  let lenA = 0, lenB = 0
  let ptrA = headA, ptrB = headB
  while (ptrA) {
    ptrA = ptrA.next
    lenA++
  }
  while (ptrB) {
    ptrB = ptrB.next
    lenB++
  }

  let dif = Math.abs(lenA - lenB)
  if (lenA > lenB) {
    while (dif--) headA = headA.next
  } else if (lenA < lenB) {
    while (dif--) headB = headB.next
  }

  while (headA && headB) {
    if (headA === headB) return headA
    headA = headA.next
    headB = headB.next
  }

  return null
};